src/metabase/upload.clj

```diff
@@ -340,29 +293,69 @@
     name
     (str schema "." name)))
 
+(defn- parse-rows
+  "Returns a lazy seq of parsed rows, given a sequence of upload types for each column.
+  Replaces empty strings with nil."
+  [col-upload-types rows]
+  (let [settings (upload-parsing/get-settings)
+        parsers  (map #(upload-parsing/upload-type->parser % settings) col-upload-types)]
+    (for [row rows]
+      (for [[value parser] (map-with-nils vector row parsers)]
+        (when (not (str/blank? value))
+          (parser value))))))
+
+(defn- remove-indices
+  "Removes the elements at the given indices from the collection. Indices is a set."
+  [indices coll]
+  (keep-indexed (fn [idx item]
+                  (when-not (contains? indices idx)
+                    item))
+                coll))
+
+(defn- indices-where
+  "Returns a lazy seq of the indices where the predicate is true."
+  [pred coll]
+  (keep-indexed (fn [idx item]
+                  (when (pred item)
+                    idx))
+                coll))
+
+(defn- auto-pk-column-indices
+  "Returns the indices of columns that have the same normalized name as [[auto-pk-column-name]]"
+  [header]
+  (set (indices-where #(= auto-pk-column-name (normalize-column-name %)) header)))
+
+(defn- rows-without-auto-pk-columns
```

Comment:
```suggestion
(defn- csv-without-auto-pk-columns
```

This deals with both header and rows, so calling it `rows` is misleading

Reply:
`csv` is misleading too, because it's kinda independent of the format because the CSV has already been parsed into rows and columns. How about `without-auto-pk-columns`? Then anything that has columns (like the header and rows) are acceptable and there's no confusion

Reply:
I renamed it to `without-auto-pk-columns` and used `header-and-rows` in the arglist

```
(defn- without-auto-pk-columns
  [header-and-rows]
  (let [header (first header-and-rows)
        auto-pk-indices (auto-pk-column-indices header)]
    (cond->> header-and-rows
      auto-pk-indices
      (map (partial remove-indices auto-pk-indices)))))
```